File: lib/cpp/math.cpp
#include <cmath>
#include <iostream>
#include <numeric>
#include <tuple>
#include <vector>


using namespace std;

// 최대공약수 (GCD)
long long gcd(long long a, long long b) {
  while (b) {
    a %= b;
    swap(a, b);
  }
  return a;
}

// 최소공배수 (LCM)
long long lcm(long long a, long long b) {
  if (a == 0 || b == 0)
    return 0;
  return (a / gcd(a, b)) * b;
}

// 확장 유클리드 (ax + by = gcd(a, b) 해 찾기)
// 반환값: {g, x, y} -> g = ax + by
tuple<long long, long long, long long> extended_gcd(long long a, long long b) {
  if (b == 0)
    return {a, 1, 0};
  auto [g, x, y] = extended_gcd(b, a % b);
  return {g, y, x - (a / b) * y};
}

// 모듈러 역원 (a^-1 mod m)
// gcd(a, m) != 1 이면 존재하지 않음 (-1 반환)
long long modInverse(long long a, long long m) {
  auto [g, x, y] = extended_gcd(a, m);
  if (g != 1)
    return -1;
  return (x % m + m) % m;
}

// 빠른 거듭제곱 (base^exp % mod)
long long power(long long base, long long exp, long long mod) {
  long long res = 1;
  base %= mod;
  while (exp > 0) {
    if (exp % 2 == 1)
      res = (res * base) % mod;
    base = (base * base) % mod;
    exp /= 2;
  }
  return res;
}

// 에라토스테네스의 체 (소수 판별)
vector<bool> sieve(int n) {
  vector<bool> is_prime(n + 1, true);
  is_prime[0] = is_prime[1] = false;
  for (int p = 2; p * p <= n; p++) {
    if (is_prime[p]) {
      for (int i = p * p; i <= n; i += p)
        is_prime[i] = false;
    }
  }
  return is_prime;
}

// 오일러 피 함수 (1~n 중 n과 서로소인 수의 개수)
long long euler_phi(long long n) {
  long long result = n;
  for (long long i = 2; i * i <= n; i++) {
    if (n % i == 0) {
      while (n % i == 0)
        n /= i;
      result -= result / i;
    }
  }
  if (n > 1)
    result -= result / n;
  return result;
}

// 조합 (nCr % mod) - 팩토리얼 미리 계산 필요
const int MAX = 1000005;
const int MOD = 1e9 + 7;
long long fact[MAX], invFact[MAX];

void precomputeFactorial() {
  fact[0] = 1;
  for (int i = 1; i < MAX; i++)
    fact[i] = (fact[i - 1] * i) % MOD;
  invFact[MAX - 1] = power(fact[MAX - 1], MOD - 2, MOD);
  for (int i = MAX - 2; i >= 0; i--)
    invFact[i] = (invFact[i + 1] * (i + 1)) % MOD;
}

long long nCr(int n, int r) {
  if (r < 0 || r > n)
    return 0;
  return fact[n] * invFact[r] % MOD * invFact[n - r] % MOD;
}

File: lib/py/math.py
import math

# 최대공약수 (GCD)
def gcd(a, b):
    while b:
        a, b = b, a % b
    return a

# 최소공배수 (LCM)
def lcm(a, b):
    if a == 0 or b == 0: return 0
    return abs(a * b) // gcd(a, b)

# 확장 유클리드 (ax + by = gcd(a, b))
# 반환: (g, x, y)
def extended_gcd(a, b):
    if b == 0:
        return a, 1, 0
    g, x, y = extended_gcd(b, a % b)
    return g, y, x - (a // b) * y

# 모듈러 역원 (a^-1 mod m)
def mod_inverse(a, m):
    g, x, y = extended_gcd(a, m)
    if g != 1:
        return None  # 역원 존재 X
    return (x % m + m) % m

# 빠른 거듭제곱 (base^exp % mod)
# Python 내장 pow(base, exp, mod) 사용 권장
def power(base, exp, mod):
    return pow(base, exp, mod)

# 에라토스테네스의 체
def sieve(n):
    is_prime = [True] * (n + 1)
    is_prime[0] = is_prime[1] = False
    for p in range(2, int(n**0.5) + 1):
        if is_prime[p]:
            for i in range(p * p, n + 1, p):
                is_prime[i] = False
    return is_prime

# 오일러 피 함수
def euler_phi(n):
    result = n
    p = 2
    while p * p <= n:
        if n % p == 0:
            while n % p == 0:
                n //= p
            result -= result // p
        p += 1
    if n > 1:
        result -= result // n
    return result

# 조합 (nCr % mod)
class Combinatorics:
    def __init__(self, n, mod):
        self.mod = mod
        self.fact = [1] * (n + 1)
        self.inv = [1] * (n + 1)
        
        for i in range(1, n + 1):
            self.fact[i] = (self.fact[i - 1] * i) % mod
            
        self.inv[n] = pow(self.fact[n], mod - 2, mod)
        for i in range(n - 1, -1, -1):
            self.inv[i] = (self.inv[i + 1] * (i + 1)) % mod
            
    def nCr(self, n, r):
        if r < 0 or r > n: return 0
        num = self.fact[n]
        den = (self.inv[r] * self.inv[n - r]) % self.mod
        return (num * den) % self.mod
